### ISSUE TITLE
Use the GGX-Smith distribution of visible normals (VNDF)

### ISSUE THREAD

**Issue by pixnblox** [materialx/MaterialX#703]
We would like to use the **GGX-Smith distribution of visible normals (VNDF)** as the default microfacet distribution in MaterialX. This provides the same final result as the standard GGX NDF, but with lower variance when sampling. In other words, it will take fewer samples to achieve the same quality, or higher quality with the same number of samples. The implementation assumes the Smith geometry term, but that is already used in MaterialX.

This would be implemented as a drop-in replacement for `mx_ggx_importance_sample_NDF` and `mx_ggx_pdf`. Specifically, this can be used for the GGX sampling currently done for directional albedo and filtered importance sampling in the MaterialX GLSL code:

https://github.com/materialx/MaterialX/blob/main/libraries/pbrlib/genglsl/lib/mx_microfacet_specular.glsl#L76
https://github.com/materialx/MaterialX/blob/main/libraries/pbrlib/genglsl/lib/mx_environment_fis.glsl#L13

The attached file contains an HLSL implementation of the paper "[Sampling the GGX Distribution of Visible Normals](http://jcgt.org/published/0007/04/01)" (2018) by Heitz, with extensive comments. This is used at Autodesk for the path tracing renderer presented at NVIDIA GTC 2021, and is donated here for use in MaterialX. It will have to be ported to GLSL, which should be straightforward. The relevant functions in this implementation are `sampleGGXSmithVNDF` and `computeGGXSmithPDF`.

[GGX-Smith BRDF.txt](https://github.com/materialx/MaterialX/files/7241177/GGX-Smith.BRDF.txt)

**Comment by pixnblox** [comment]
I discussed this a bit with @jstone-lucasfilm and he may have thoughts on how this can proceed.

**Comment by jstone-lucasfilm** [comment]
This approach looks very promising, thanks @pixnblox, and you can see initial results for our GGX directional albedo computation in pull request #706.

**Comment by pixnblox** [comment]
@jstone-lucasfilm: Thanks for following up on this, and I am glad it is working out!

Do you expect to try it at the other place GGX sampling is performed, for filtered importance sampling?

https://github.com/materialx/MaterialX/blob/main/libraries/pbrlib/genglsl/lib/mx_environment_fis.glsl#L30

**Comment by jstone-lucasfilm** [comment]
@pixnblox Yes, I'm really interested in testing that as a next step, once this initial change has been merged.  In addition to updating the sampling function, we'd need to update all references to its PDF, which is used as an input to the LOD selection function for filtered importance sampling.

**Comment by jstone-lucasfilm** [comment]
I've done some initial investigation into the use of VNDF sampling for FIS, and wanted to report the results for those who are interested.

Although the VNDF-sampled version of FIS does indeed converge more rapidly than the NDF version, it comes with its own visual artifacts, noticeable as a warping of the lighting in regions of the screen where the view vector and surface normal are nearly parallel.  These artifacts step from the logic used to generate a tangent frame about the view vector, which (by design) does not produce smooth results in regions where the view vector and surface normal are aligned:

https://github.com/materialx/MaterialX/blob/main/libraries/pbrlib/genglsl/lib/mx_microfacet_specular.glsl#L39

It may be possible to address these issues by designing a different approach for computing the projected area of the hemisphere in VNDF sampling, but this would be a novel (though interesting) research project.

For now I'll close out this issue, since we've successfully integrated VNDF sampling into the computation of GGX directional albedo, where its benefits were clear and unambiguous.

**Comment by pixnblox** [comment]
This is now fully implemented (and with a performance improvement) in PR #1390. Thanks, Jonathan!
